src/lib/loaders/batchSaleArtworkLoader.ts

```diff
@@ -0,0 +1,59 @@
+import DataLoader from "dataloader"
+
+// Take the list of params that the loader is called with, tuples of `artworkId` and `saleId`.
+// Group them by saleId, and return a dictionary where the the value for each record
+// is the list of artwork ids for that sale.
+const groupByParams = (
+  params: { artworkId: string; saleId: string }[]
+): Record<string, string[]> => {
+  const groupedParams = params.reduce((acc, { saleId, artworkId }) => {
+    if (!acc[saleId]) {
+      acc[saleId] = []
+    }
+    acc[saleId].push(artworkId)
+    return acc
+  }, {})
+
+  return groupedParams
+}
+
+export const createBatchSaleArtworkLoader = (multipleLoader) => {
+  const dataLoader = new DataLoader(
+    async (idWithParamsList: { artworkId: string; saleId: string }[]) => {
+      const groupedParams = groupByParams(idWithParamsList)
+
+      // Dictionary to store resolved data for each saleId.
+      // The value stored is a list of fetched sale artworks for that sale.
+      const data: Record<string, { artwork: { _id: string } }[]> = {}
+
+      // Fetch data for each saleId using a list endpoint.
+      for (const saleId in groupedParams) {
```

Comment:
Can we do Promise.all here?

Reply:
Good call, this awaits serially per sale (which for auction page/rail won't matter as it's all the same sale - but still).